### PR TITLE
Add action to display leaderboard in iframe

### PR DESCRIPTION
**Pull request by AlexandreSi**

### CODE REVIEW

File: Extensions/Leaderboards/JsExtension.js
Comment (AlexandreSi) [reply Extensions/Leaderboards/JsExtension.js:133]: Well, they are a bit different : if you open X leaderboards, you don't open X views: the first leaderboard opens the view, the followings just change the href. So the close leaderboard view seems ok to me.
BUT for sure, it may be not clear for our users. Maybe I can add "If a leaderboard is already open, it will replace it in the view"? in the action `DisplayLeaderboard`?

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [inline Extensions/Leaderboards/JsExtension.js:119]: No period at the end of an events sheet sentence

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:150]:
```suggestion
              "The div element covering the game couldn't be found, the leaderboard cannot be displayed."
```

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:170]: Is `window as Window` working?

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [reply Extensions/Leaderboards/JsExtension.js:118]: Agree with the other comment saying that we can also say "If a leaderboard was already displayed on top of the game, the new leaderboard will replace it"

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [inline Extensions/Leaderboards/JsExtension.js:134]:
```suggestion
        _('Close the leaderboard currently displayed on top of the game.'),
```

To be a bit more precise?

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (AlexandreSi) [reply Extensions/Leaderboards/leaderboardstools.ts:170]: No, TS doesn't seem to be able to cast it

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [inline Extensions/Leaderboards/JsExtension.js:129]: Let's add a condition "LeaderboardViewErrored" ("The leaderboard view had an error when opening") which returns true when we detect the leaderboard view could not be open, so that the game can react to this :)

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:142]: For the loading error handling:
- Set back the `_leaderboardViewIframeErrored` to false here.
- In theory the event listener for "error" and "load" will then set it to true (or to false again).

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:144]: For the loading error handling:
- Set back the `_leaderboardViewIframeErrored` to false here.

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:152]: For the loading error handling:
- Set back the `_leaderboardViewIframeErrored` to true here.

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:156]: For the loading error handling:
- Before setting `src`, add some `addEventListener` to listen to `load` and `error`. 
- In case of error, log it and set  the `_leaderboardViewIframeErrored` to true, then close the leaderboard view (the iframe could not load, we don't want to soft lock the game on a blank screen).
- In the "load" event: Set it to false, for robustness/safety sake.

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:183]: ~~For the loading error handling:~~
- ~~Reset `_leaderboardViewIframeErrored` to false at the beginning of this method~~

Actually don't do it, because it's important that we can close the leaderboard in case of an error, and still be able to check from the game conditions that an error happened.

The error flag should be reset to false only when we open a new leaderboard view.

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [reply Extensions/Leaderboards/JsExtension.js:129]: (see other comments about "loading error handling")

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [inline Extensions/Leaderboards/JsExtension.js:130]: I wonder if we also should have a condiiton "LeaderViewOpened" which means the iframe was created and is displayed on top of the game (even if it is still loading). This way we have:

- An action to open the leaderboard view.
- We have a condition to know if it's opened (useful to pause your game or do something in the background).
- We have a condition to check if it errored (so you can at least display "oops, try again later") and could not be displayed.
- We have an action to close the leaderboard (though I guess in the majority of cases, it's the player that will close it).

**Review by 4ian — COMMENTED** [review]
Good to go once we've bulletproofed how the leaderboard view react in case of error.
This can be:
- A loading error.
- "Worse", an error like a 404 which displays the wrong page. I'm not sure about this one. Maybe we could try to detect in the iframe the presence of the close button or some DOM element with a well known id. If it's not found 2 seconds after the loading of the iframe is done => we close the leaderboard view considering it errored.

This last case is more tricky (we have to ensure that as a parent we can read the content of the iframe... which is not guaranteed because browser could put restrictions => need more investigation).

Let's at least make the loading robust in case of loading error (which you can test by entering a wrong url I think).

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:285]: These .catch after a then will catch errors launched by `checkLeaderboardAvailability`, but also by the entire body of the `then`. If it's not meant to do that, consider using then 2nd argument instead: `.then(() => {...}, (error) => {...})`.
(Catching everything seems like a good idea, but could also hide real errors like syntax errors, unhandled exception and generally it's better to let them just go rather than hiding them behind an error message that is unrelated and confusing)

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:259]: Very useful comment, thanks!

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:31]: animate supports started with `iOS 13.1`... I think it's good but we could almost put a try/catch around it to allow this to work (without animation) without crashing the game on iOS 12/13 (even if it's <2% of iOS usage).

Comment (4ian) [reply Extensions/Leaderboards/leaderboardstools.ts:31]: Also, will this consume CPU even if not rendered? 🤔

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:13]: Interesting, why the choice of `NodeJS.Timeout`?

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:224]: Add a TODO when you leave development urls ;) Easier to spot before merging

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (4ian) [inline Extensions/Leaderboards/leaderboardstools.ts:220]: Last (?) edge case: if I call displayLeaderboard multiple times quickly with different leaderboard ids.
Like, `displayLeaderboard(.., 1)`, `displayLeaderboard(.., 2)` and finally `displayLeaderboard(.., 3)`.

There is small, but non zero, chance that 1 will take time and the response (of `checkLeaderboardAvailability`) will arrive after the response for 2 and 3.

So you will end up displaying the leaderboard 1 instead of the last one that was requested - which will be surprising and surely considered as a bug/strange behavior.

There is no simple way to avoid this - ideally we would have something to "cancel" promises that are "in flight" and stop execution of anything left.
As this is not easily doable in JS, you can still store the `_requestedLeaderboardId`. You set it directly at the beginning of the call - so we are assured this is up to date with the latest call.
When you received the response of checkLeaderboardAvailability, you check `if (_requestedLeaderboardId !== leaderboardId)`. If that's true, you are getting a response for something outdated. You simply bail out (and don't error - you just don't care about the response anymore).

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [inline Extensions/Leaderboards/JsExtension.js:95]: Nitpicking: we took the convention to start all conditions and actions by a capital (PascalCase)

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [inline Extensions/Leaderboards/JsExtension.js:97]:
```suggestion
        _('Leaderboard display has errored'),
```

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [inline Extensions/Leaderboards/JsExtension.js:99]:
```suggestion
        _('Leaderboard display has errored'),
```

File: Extensions/Leaderboards/JsExtension.js
Comment (4ian) [inline Extensions/Leaderboards/JsExtension.js:113]: I would even call this "Leaderboard display has loaded and is displayed on screen"

**Review by 4ian — APPROVED** [review]
We're close! I think this will be ready to go after this round of comments!

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (AlexandreSi) [reply Extensions/Leaderboards/leaderboardstools.ts:31]: > animate supports started with iOS 13.1... I think it's good but we could almost put a try/catch around it to allow this to work (without animation) without crashing the game on iOS 12/13 (even if it's <2% of iOS usage).

Good idea, I'll add it

> Also, will this consume CPU even if not rendered? 🤔

Good question, I'll check

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (AlexandreSi) [reply Extensions/Leaderboards/leaderboardstools.ts:13]: That's the type returned by `setTimeout` 😅

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (AlexandreSi) [reply Extensions/Leaderboards/leaderboardstools.ts:224]: 😓

File: Extensions/Leaderboards/leaderboardstools.ts
Comment (AlexandreSi) [reply Extensions/Leaderboards/leaderboardstools.ts:31]: I couldn't find something on the internet explaining how Animate would behave on a node that is not in the DOM.
I made a test on a dedicated HTML file. Here are the performance analysis by Chrome DevTools:
- First case: Loader is inserted in the div at start, then animated at 1.5s.

image

- Second case: Loader is inserted in the div at 3s, bu still animated at 1.5s

image

The animation seems to not require CPU when the element is not displayed: CPU line is hardly visible but:
- in the first case, we can see a little "peak" at 1.5s 
- in the second case, there seems to be a small CPU Usage when the element appears at 3s but none when the animation starts at 1.5s

Comment (AlexandreSi) [reply Extensions/Leaderboards/leaderboardstools.ts:31]: Another point is that I found a few articles claiming that infinite CSS animations can consume a lot of CPU. But the Animate API doesn't seem to inject CSS on the element.

Comment (4ian) [reply Extensions/Leaderboards/leaderboardstools.ts]: > The animation seems to not require CPU when the element is not displayed

Ok! I'm a bit paranoid on having any DOM work that would still the browser CPU time from the game and that would make the browser not 100% dedicated on running the game frames.

**Pull request merged**
